File: leetcode/java/39.Combination Sum.java
import java.util.*;
import java.io.*;

class Main{

   static PrintStream ps = System.out;
  
   public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if(candidates == null || candidates.length == 0) return result;
        ArrayList<Integer> current = new ArrayList<Integer>();
        Arrays.sort(candidates);
        combinationSum(candidates, target, 0, current, result);
        return result;
    }

    public static void combinationSum(int[] candidates, int target, int j, ArrayList<Integer> curr, List<List<Integer>> result){
        if(target == 0){
            ArrayList<Integer> temp = new ArrayList<Integer>(curr);
            result.add(temp);
            return;
        }
        // if target is zero add curr to result & clear it
        for(int i=j; i<candidates.length; i++){
            if(target < candidates[i])
                return;
            curr.add(candidates[i]);
            combinationSum(candidates, target - candidates[i], i, curr, result);
            curr.remove(curr.size()-1);
       }
   }

  public static void main(String ...st){
    combinationSum(new int[]{2,3,6,7},7).stream().forEach(x->{
       ps.println(x);
     });
  }

}

File: leetcode/kotlin/39.Combination Sum.kt
import java.util.*

internal object Main {
    private var ps = System.out
    private fun combinationSum(candidates: IntArray?, target: Int): List<List<Int>> {
        val result: MutableList<List<Int>> = ArrayList()
        if (candidates == null || candidates.isEmpty()) return result
        val current = ArrayList<Int>()
        Arrays.sort(candidates)
        combinationSum(candidates, target, 0, current, result)
        return result
    }

    private fun combinationSum(
        candidates: IntArray,
        target: Int,
        j: Int,
        curr: ArrayList<Int>,
        result: MutableList<List<Int>>
    ) {
        if (target == 0) {
            val temp = ArrayList(curr)
            result.add(temp)
            return
        }
        // if target is zero add curr to result & clear it
        for (i in j until candidates.size) {
            if (target < candidates[i]) return
            curr.add(candidates[i])
            combinationSum(candidates, target - candidates[i], i, curr, result)
            curr.removeAt(curr.size - 1)
        }
    }

    @JvmStatic
    fun main(st: Array<String>) {
        combinationSum(intArrayOf(2, 3, 6, 7), 7).stream().forEach { x: List<Int>? ->
            ps.println(
                x
            )
        }
    }
}